include/aws/common/rw_lock.h

```diff
@@ -21,29 +21,40 @@
 /* NOTE: Do not use this macro before including Windows.h */
 #    define AWSSRW_TO_WINDOWS(pCV) (PSRWLOCK) pCV
 #else
-#    include <pthread.h>
+#    include <aws/common/atomics.h>
+#    include <aws/common/mutex.h>
+#    include <aws/common/semaphore.h>
 #endif
 
 struct aws_rw_lock {
 #ifdef _WIN32
     void *lock_handle;
 #else
-    pthread_rwlock_t lock_handle;
+    struct aws_atomic_var readers;
+    struct aws_atomic_var holdouts;
+    struct aws_semaphore reader_sem;
+    struct aws_semaphore writer_sem;
+    struct aws_mutex writer_lock;
 #endif
 };
 
 #ifdef _WIN32
 #    define AWS_RW_LOCK_INIT                                                                                           \
         { .lock_handle = NULL }
 #else
+/* -1 ^ (1ull << (sizeof(size_t) * 8 - 1)): SSIZE_MAX but without guaranteeing SSIZE_MAX exists. */
 #    define AWS_RW_LOCK_INIT                                                                                           \
-        { .lock_handle = PTHREAD_RWLOCK_INITIALIZER }
+        {                                                                                                              \
+            .readers = AWS_ATOMIC_INIT_INT(0), .holdouts = AWS_ATOMIC_INIT_INT(0),                                     \
+            .reader_sem = AWS_SEMAPHORE_INIT(0, -1 ^ (1ull << (sizeof(size_t) * 8 - 1))),                              \
```

Comment:
This isn't true, because I'm always casting the value to a signed type, and we need to be able to subtract max_readers without underflowing the value. The max readers must NOT be the unsigned max, it must be the signed max or else we would violate this requirement. You can verify this by changing it to SIZE_MAX locally, where you'll notice the tests deadlock.

I can just branch on SIZE_MAX == UINT32/64_MAX though, that's a good idea.

Reply:
Do we ever read the semaphore's max count? I agree that changing `s_max_readers` to SIZE_MAX would break, but I don't think the semaphore limit should impact anything.

Reply:
They do need to be the same, because otherwise rlock would let more readers in than max readers, so when wlock subtracts max readers, the number would still be positive, breaking all sorts of logic.